src/background/services/deduplicator.test.ts

```diff
@@ -0,0 +1,192 @@
+      const resultPromises = [dedupedFn(), dedupedFn(), dedupedFn()];
+
+      // advance timers to trigger the async function
+      jest.runAllTimers();
+
+      const results = await Promise.all(resultPromises);
+
+      expect(results[0]).toBe(returnValue);
+      expect(results[1]).toBe(returnValue);
+      expect(results[2]).toBe(returnValue);
+      expect(fn).toHaveBeenCalledTimes(1);
+    });
+
+    it('should support different function signatures', async () => {
+      const fn1 = createAsyncFn({ returnValue: 'hello', timeout: 100 });
+      const fn2 = createAsyncFn({ returnValue: 'world', timeout: 400 });
+      const dedupedFn1 = deduplicatorService.dedupe(fn1, {
+        cacheFnArgs: true,
+      });
+      const dedupedFn2 = deduplicatorService.dedupe(fn2, {
+        cacheFnArgs: true,
+      });
+      // createAsyncFn function returns an anonymous function, which is created using the new Promise constructor.
+      // it does not have a name property, so it does NOT have a key for dedup
+      const resultPromises = [dedupedFn1('arg1'), dedupedFn2('arg2')];
+
+      jest.runAllTimers();
+
+      const [result1, result2] = await Promise.all(resultPromises);
+
+      expect(result1).toBe('hello');
+      expect(result2).toBe('world');
+      expect(fn1).toHaveBeenCalledTimes(1);
+      expect(fn2).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe('Cache Arguments Configuration', () => {
+    it('should differentiate calls with different arguments when cacheFnArgs is true', async () => {
+      const returnValue = 2n;
+      const fn = createAsyncFn({ returnValue });
+      const dedupedFn = deduplicatorService.dedupe(fn, {
+        cacheFnArgs: true,
+      });
+
+      dedupedFn('arg1');
+      dedupedFn('arg2');
+      jest.runAllTimers();
+
+      expect(fn).toHaveBeenCalledTimes(2);
+      expect(fn).toHaveBeenCalledWith('arg1');
+      expect(fn).toHaveBeenCalledWith('arg2');
+    });
+
+    it('should cache calls with the same arguments when cacheFnArgs is true', async () => {
+      const returnValue = 2n;
+      const fn = createAsyncFn({ returnValue });
+      const dedupedFn = deduplicatorService.dedupe(fn, {
+        cacheFnArgs: true,
+      });
+
+      const resultPromises = [dedupedFn(1, 2), dedupedFn(1, 2)];
+      jest.runAllTimers();
+      const [result1, result2] = await Promise.all(resultPromises);
+
+      expect(result1).toBe(returnValue);
+      expect(result2).toBe(returnValue);
+      expect(fn).toHaveBeenCalledTimes(1);
+    });
+  });
+
+  describe('Error Handling, cacheRejections', () => {
+    it('should not cache rejections by default', async () => {
+      const fn = createAsyncFn({
+        returnValue: 2n,
+        shouldReject: true,
+        timeout: 500,
+      });
+      const dedupedFn = deduplicatorService.dedupe(fn);
+
+      const resultPromises = [dedupedFn(1, 2), dedupedFn(1, 2)];
+      jest.runAllTimers();
+      const [result1, result2] = await Promise.all(resultPromises);
```

Comment:
`await Promise.all` will throw before our `expect`:

```suggestion
      const [result1, result2] = [dedupedFn(1, 2), dedupedFn(1, 2)];
      jest.runAllTimers();
```